File: solutions/problems/dynamicProgramming/houseRobber.ts
/**
 * Approach: rolling DP. prev1 = best up to the previous house, prev2 =
 * best up to two houses back. At each house, take the better of skipping
 * it (prev1) or robbing it (its value + prev2). Time: O(n). Space: O(1).
 */

export function rob(nums: number[]): number {
  let prev2 = 0;
  let prev1 = 0;
  for (const num of nums) {
    [prev2, prev1] = [prev1, Math.max(prev1, prev2 + num)];
  }
  return prev1;
}

File: problems/dynamic_programming/house_robber.py
"""
LeetCode 198. House Robber

Houses are in a row, each with some money. You can't rob two adjacent
houses. Maximize the total amount robbed.

Approach: at each house, decide to skip it (carry forward the best from
before) or rob it (its value + best up to two houses back). Track just the
last two running bests (rolling DP) instead of a full array.

Time:  O(n)
Space: O(1)
"""

from typing import List


def rob(nums: List[int]) -> int:
    prev2, prev1 = 0, 0  # best up to i-2, best up to i-1
    for num in nums:
        prev2, prev1 = prev1, max(prev1, prev2 + num)
    return prev1


if __name__ == "__main__":
    assert rob([1, 2, 3, 1]) == 4
    assert rob([2, 7, 9, 3, 1]) == 12
    assert rob([]) == 0
    assert rob([5]) == 5
    print("All tests passed.")
